sdk/src/types/mod.rs

```diff
@@ -3,8 +3,143 @@
 
 //! Common types required by nodes and clients APIs like blocks, responses and DTOs.
 
+use core::ops::Deref;
+
+use self::block::protocol::ProtocolParameters;
+
 #[cfg(feature = "std")]
 extern crate std;
 
 pub mod api;
 pub mod block;
+
+/// Borrowed or Owned. Useful for generic impls and UX.
+#[derive(Clone, Debug)]
+pub enum Boo<'a, T> {
+    Borrowed(&'a T),
+    Owned(T),
+}
+
+impl<'a, T> Boo<'a, T> {
+    pub fn into_owned(self) -> T
+    where
+        T: Clone,
+    {
+        match self {
+            Boo::Borrowed(b) => b.clone(),
+            Boo::Owned(o) => o,
+        }
+    }
+}
+
+impl<'a, T> AsRef<T> for Boo<'a, T> {
+    fn as_ref(&self) -> &T {
+        self.deref()
+    }
+}
+
+impl<'a, T> Deref for Boo<'a, T> {
+    type Target = T;
+
+    fn deref(&self) -> &Self::Target {
+        match *self {
+            Boo::Borrowed(b) => b,
+            Boo::Owned(ref o) => o,
+        }
+    }
+}
+
+impl<'a, T> From<T> for Boo<'a, T> {
+    fn from(value: T) -> Self {
+        Self::Owned(value)
+    }
+}
+
+impl<'a, T> From<&'a T> for Boo<'a, T> {
+    fn from(value: &'a T) -> Self {
+        Self::Borrowed(value)
+    }
+}
+
+impl<'a, T> From<&'a Boo<'a, T>> for Boo<'a, T> {
+    fn from(value: &'a Boo<'a, T>) -> Self {
+        match *value {
+            Self::Borrowed(b) => Self::Borrowed(b),
+            Self::Owned(ref o) => Self::Borrowed(o),
+        }
+    }
+}
+
+#[derive(Clone, Default, Debug)]
+pub struct ValidationParams<'a> {
+    protocol_parameters: Option<Boo<'a, ProtocolParameters>>,
+    token_supply: Option<u64>,
```

Comment:
Does it still make sense to have the token_supply as a separated field? It's also in the ProtocolParameters

Reply:
Yes, but this way you can provide only the token supply if that's all you want to verify.

Reply:
Yeah I know, my question was just if we still use it, so it makes sense to have
Didn't check when I asked, but did now and we still use it in many places so all good